Update ``lightning.tensor`` docs (#909)
PennyLaneAI/pennylane-lightning: doc/lightning_tensor/device.rst

    - ``cutoff`` (singular value truncation threshold) defaults to ``0`` .
    - ``cutoff_mode`` (singular value truncation mode) defaults to ``abs`` , considering the absolute values of the singular values; Alternatively, users can opt to set ``cutoff_mode`` to ``rel`` to consider the relative values of the singular values.
Note that the ``cutensornet`` will automatically determine the reduced extent of the bond dimension based on the lowest among the multiple truncation cutoffs (``max_bond_dim``, ``cutoff-abs`` and ``cutoff-rel``). For more details on how the ``cutoff`` works, please check it out the `cuQuantum documentation <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/types.html#cutensornettensorsvdconfigattributes-t>`__.

The ``lightning.tensor`` device dispatches all operations to be performed on a CUDA-capable GPU of generation SM 7.0 (Volta)
and greater. This device supports both exact and finite shots measurements. Currently, the supported differentiation methods are parameter-shift and finite-diff. Note that the MPS backend of lightning.tensor supports multi-wire gates via Matrix Product Operators (MPO).

> The ``lightning.tensor`` device is designed for expectation value calculations. Measurements of ``qml.probs()`` or ``qml.state()`` return dense vectors of dimension :math:`2^{n_\text{qubits}}`, so they should only be used for small systems. When using finite shots, ``qml.probs()`` can be used to get individual samples.

Reviewer: > Shots-based calculation does not require an (2**n_wires*8 bytes) additional GPU memoery

Can you please help me understand this a better - why does shot-based calculation use less memory than the exact case?

> The ``lightning.tensor`` device is designed for expectation value calculations. Measurements of ``qml.probs()`` or ``qml.state()`` return dense vectors of dimension :math:`2^{n_\text{qubits}}`, so they should only be used for small systems. When using finite shots, ``qml.probs()`` can be used to get individual samples.

Author: To make it more clear, less memory requirement for shot measurements here means that we don't have to allocate GPU memory (`2**num_wires`) in our C++ impelementations to store the full/subset state, which is required for the prob calculations. Shot-measurements return the generated samples to CPU memory directly and the required memory for the generate samples is (`num_shots*num_qubits*8`) bytes. Hope it makes it easier to understand. We can schedule a call if needed.

> .. note:: ``qml.Hermitian`` is currently only supported for single wires. You can use ``qml.pauli_decompose`` on smaller matrices to obtain a compatible Pauli decomposition in the meantime. As a result, it is advisable to disable ``new_opmath`` for the ``lightning.tensor`` device to prevent the implicit conversion of Pauli words to multi-wire Hermitian observables. This limitation will be addressed once multi-wires Hermitian observables can be supported with ``cutensornet``.

Reviewer: I still dont understand this limitation. Are we restricted specifically for ``qml.Hermitian`` for more than one wire, or is it that generally we dont support observables with operators on more than one wire (i.e. also not supporting things like `qml.X(0) @ qml.X(1)`? From your previous comment here, it sounded like the latter is the case, but this here again reads like it is only a restriction for ``qml.Hermitian`` - which one is it?

And more generally what is the root of this restriction? How come this works with legacy opmath and not new opmath? Can we fix this before release? I would deem this high priority, as such a suggestion and limitation is quite severe.

Author: This ticket is added to Q4!

The default setup for the MPS tensor network approximation is:
    - ``max_bond_dim`` (maximum bond dimension) defaults to ``128`` .
    - ``cutoff`` (singular value truncation threshold) defaults to ``0`` .
    - ``cutoff_mode`` (singular value truncation mode) defaults to ``abs`` , considering the absolute values of the singular values; Alternatively, users can opt to set ``cutoff_mode`` to ``rel`` to consider the relative values of the singular values.

> Note that the ``cutensornet`` will automatically determin the reduced extent of the bond dimension based on the lowest among the multiple truncation cutoffs (``max_bond_dim``, ``cutoff-abs`` and ``cutoff-rel``). For more details on how the ``cutoff`` works, please check it out the `cuQuantum documentation <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/types.html#cutensornettensorsvdconfigattributes-t>`__. All swap and decomposition operations in MPS-MPO multiplication will follow the same constraints set by SVD configurations (``cutoff``) and target extents setups. For more details, please refer to the `cuQuantum documentation <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/types.html#cutensornetstatempoapplication-t>`__.

Reviewer: > All swap and decomposition operations in MPS-MPO multiplication will follow the same constraints set by SVD configurations (``cutoff``) and target extents setups.

Is this really true? Perhaps I misunderstood your reply [here](https://github.com/PennyLaneAI/pennylane-lightning/pull/859#discussion_r1761592785) but to me this sounded like the cutoff truncation is not performed on the MPO? (just the MPS after the truncation)

> Note that the ``cutensornet`` will automatically determin the reduced extent of the bond dimension based on the lowest among the multiple truncation cutoffs (``max_bond_dim``, ``cutoff-abs`` and ``cutoff-rel``). For more details on how the ``cutoff`` works, please check it out the `cuQuantum documentation <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/types.html#cutensornettensorsvdconfigattributes-t>`__. All swap and decomposition operations in MPS-MPO multiplication will follow the same constraints set by SVD configurations (``cutoff``) and target extents setups. For more details, please refer to the `cuQuantum documentation <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/types.html#cutensornetstatempoapplication-t>`__.

Author: Hey @Qottmann , I haven't had a chance to check how the `cutoff` works under the hood.  The statement above is based on the cuquantum docs below.

>When multiple truncation cutoffs (CUTENSORNET_TENSOR_SVD_CONFIG_ABS_CUTOFF, CUTENSORNET_TENSOR_SVD_CONFIG_REL_CUTOFF, CUTENSORNET_TENSOR_SVD_CONFIG_DISCARDED_WEIGHT_CUTOFF) or maximal extent in input [cutensornetTensorDescriptor_t](https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/types.html#types_8h_1a35bd3aace8877207eae09420ac9cdc9e) are specified. The runtime reduced extent will be determined as the lowest among all.

> Note that the ``cutensornet`` will automatically determin the reduced extent of the bond dimension based on the lowest among the multiple truncation cutoffs (``max_bond_dim``, ``cutoff-abs`` and ``cutoff-rel``). For more details on how the ``cutoff`` works, please check it out the `cuQuantum documentation <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/types.html#cutensornettensorsvdconfigattributes-t>`__. All swap and decomposition operations in MPS-MPO multiplication will follow the same constraints set by SVD configurations (``cutoff``) and target extents setups. For more details, please refer to the `cuQuantum documentation <https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/types.html#cutensornetstatempoapplication-t>`__.

Reviewer: > I haven't had a chance to check how the cutoff works under the hood. 

That's fine, I'd just remove that sentence then:

> All swap and decomposition operations in MPS-MPO multiplication will follow the same constraints set by SVD configurations (``cutoff``) and target extents setups.

The ``lightning.tensor`` device dispatches all operations to be performed on a CUDA-capable GPU of generation SM 7.0 (Volta)

> and greater. This device supports both analytical and finite shots measurements. Currently, the supported differentiation methods are parameter-shift and finite-diff. Note that MPS backend of lightning.tensor supports multi-wires gate via Matrix Product Operators (MPO). Disable the new arithmetic operator dunders is recommended since only 1-wire Hermitian is supported by lightning.tensor. It is recommended to use the shot measurements for the lightning.tensor device. The calculation of analytical probabilities and variances may result in excessive memory usage for large-scale quantum simulations or become impractical due to prohibitive computational costs. It's also recommended to disable new_opmath for the lightning.tensor device since lightning.tensor only supports 1-wire Hermitian observables.

Reviewer: ```suggestion
and greater. This device supports both exact and finite shots measurements. Currently, the supported differentiation methods are parameter-shift and finite-diff. Note that the MPS backend of lightning.tensor supports multi-wire gates via Matrix Product Operators (MPO).

Note that measurements of ``qml.probs()`` or ``qml.state()`` return dense vectors of dimension :math:`2^{n_\text{qubits}}`, so they should only be used for small systems. When using finite shots, ``qml.probs()`` can be used to get individual samples.

.. note:: ``qml.Hermitian`` is currently only supported for single wires. You can use ``qml.pauli_decompose`` on smaller matrices to obtain a compatible Pauli decomposition in the meantime.
```

To me "analytic" is orthogonal to numerical, do you mean "exact"? Alternatively, you can say "infinite shots" here, specifically.

I think it would be unwise to recommend people to disable new opmath - but perhaps I am misunderstanding the current limitations?

The comments about analytic probabilities is confusing, it sounds like finite shots is better / faster in general. I tried to rephrase it a bit.

Reviewer: > It is recommended to use the shot measurements for the lightning.tensor device.

Do you really mean generally? Is this what you are observing? This sounds very strange 🤔

Author: Nice one!

Some tips on the usage of the ``lightning.tensor`` device:
    - ``lightning.tensor`` performs better for the larger bond dimensions in MPS calculations, see the ``Approximate Tensor Network Methods`` section in the `cuQuantum SDK <https://developer.nvidia.com/cuquantum-sdk>`__.
    - The ``lightning.tensor`` device is recommended for large-scale quantum simulations. For small-scale simulations, the cost of SVD decomposition and MPS-MPO multiplication might outweigh than the benifits of MPS.

>     - It is recommended to use shot-based ``probs()`` measurements for the ``lightning.tensor`` device when the number of target/projected wires is large. The analytical calculation of ``prob()`` target at ``30+`` wires can lead to excessive memory usage. The analytical calculation of ``prob()`` that targets at small number of subset wires but projects a larger number of wires could become impractical due to high computational costs for large-scale quantum simulations. If the number of projected wires is ``n``, ``2**n`` times of ``cutensornet`` API calls should be made to get the analytical ``probs()``, which could be prohibitive if ``n`` is large. 

Reviewer: see recommendation above. if you want to repeat it, try to shorten it a bit

Author: You suggestions above are accepted and don't have to repeat it here.

    - The analytical calculation of ``var()`` measurements could be impractical for Hamiltonian observables with many terms. The computational cost of current ``var`` implementation scales with ``n**2``, where ``n`` is the number of terms in the Hamiltonian observable.

>     - It is advisable to disable ``new_opmath`` for the ``lightning.tensor`` device, as the device only supports 1-wire Hermitian observables.

Reviewer: are you referring to supporting only single wire `qml.Hermitian` or more generally only supporting Hermitian operators with one wire, i.e. `qml.X(0) @ qml.Y(1)` is not supported?

Reviewer: Oh, that is indeed quite severe 😬 
And with opmath disabled, this works? Where is this coming from?

> The ``lightning.tensor`` device dispatches all operations to be performed on a CUDA-capable GPU of generation SM 7.0 (Volta)

Reviewer: > ``cutoff`` (singular value truncation threshold) defaults to ``0`` .

We are currently not actually using this, right? Perhaps worth a small comment that this is actually not supported yet

Reviewer: Or was it just for MPOs where this isnt used?

Reviewer: > ``cutoff_mode`` (singular value truncation mode) defaults to ``abs`` , considering the absolute values of the singular values; Alternatively, users can opt to set ``cutoff_mode`` to ``rel`` to consider the relative values of the singular values. 

Singular values are always non-negative and real, why do we need this at all?

Author: > >cutoff (singular value truncation threshold) defaults to 0 .

> We are currently not actually using this, right? Perhaps worth a small comment that this is actually not supported yet

``cutoff`` is supported by the ``lightning.tensor`` device. A short description on how the ``cutensornet`` backend reduce the bond extent dimensions is added.

Author: > Or was it just for MPOs where this isnt used?
It is for both `MPS` and `MPO` in the `cutensornet` context as you can find out the details in the newly added links.

Author: > > ``cutoff_mode`` (singular value truncation mode) defaults to ``abs`` , considering the absolute values of the singular values; Alternatively, users can opt to set ``cutoff_mode`` to ``rel`` to consider the relative values of the singular values. 

> Singular values are always non-negative and real, why do we need this at all?

This is a feature offered by ``cutensornet``. You can get more info [here](https://docs.nvidia.com/cuda/cuquantum/latest/cutensornet/api/types.html#cutensornettensorsvdconfigattributes-t).

and greater.

> Some tips on the usage of the ``lightning.tensor`` device:

Reviewer 2: Please integrate these points here
> This device supports both analytical and finite shots measurements. Currently, the supported differentiation methods are ``parameter-shift`` and ``finite-diff``. Note that MPS backend of ``lightning.tensor`` supports multi-wires gate via Matrix Product Operators (MPO). Disable the new arithmetic operator dunders is recommended since only 1-wire Hermitian is supported by ``lightning.tensor``. It is recommended to use the shot measurements for the ``lightning.tensor`` device. The calculation of analytical probabilities and variances may result in excessive memory usage for large-scale quantum simulations or become impractical due to prohibitive computational costs. It's also recommended to disable ``new_opmath`` for the ``lightning.tensor`` device since ``lightning.tensor`` only supports 1-wire Hermitian observables.

>     - ``lightning.tensor`` performs better for the maximum bond dimension MPS calculation. Given the inherent parallelism of GPUs, simulations with intensive parallel computation, such as those with larger maximum bond dimensions, stand to gain the most from the computational power offered by GPU and those simulations can benifit from the ``lightning.tensor`` device.  It's worth noting that if the bond dimension used in the simulation is small, the ``lightning.tensor`` device with ``MPS`` running a GPU may perform slower compared to a ``default.tensor`` device with ``MPS`` running on a CPU. For more details on how bond dimension affects the simulation performance, please refer to the ``Approximate Tensor Network Methods`` section in the `cuQuantum SDK <https://developer.nvidia.com/cuquantum-sdk>`__.

Reviewer 2: ```suggestion
    - ``lightning.tensor`` performs better for the maximum bond dimension MPS calculation. Given the inherent parallelism of GPUs, simulations with intensive parallel computation, such as those with larger maximum bond dimensions, stand to gain the most from the computational power offered by GPU and those simulations can benefit from the ``lightning.tensor`` device.  It's worth noting that if the bond dimension used in the simulation is small, the ``lightning.tensor`` device with ``MPS`` running a GPU may perform slower compared to a ``default.tensor`` device with ``MPS`` running on a CPU. For more details on how bond dimension affects the simulation performance, please refer to the ``Approximate Tensor Network Methods`` section in the `cuQuantum SDK <https://developer.nvidia.com/cuquantum-sdk>`__.
```

Reviewer 2: Just a typo

Reviewer: Can we shorten this?

```suggestion
    - ``lightning.tensor`` performs better for the larger bond dimensions in MPS calculations, see the ``Approximate Tensor Network Methods`` section in the `cuQuantum SDK <https://developer.nvidia.com/cuquantum-sdk>`__.
```

    - The ``lightning.tensor`` device is optimized for large-scale quantum simulations. For small-scale quantum simulations, the overhead of transferring data between the CPU and GPU may outweigh the benefits of GPU acceleration.

>     - For the ``lightning.tensor`` device, it is recommended to use shot-based ``probs()`` measurements. The analytical calculation of ``prob()`` can lead to excessive memory usage or become impractical due to high computational costs for large-scale quantum simulations.

Reviewer: This really confuses me, can you elaborate what you are recommending here and what is the underlying issue?

Author: > The ``lightning.tensor`` device is recommended for large-scale quantum simulations. For small-scale simulations, the cost of SVD decomposition and MPS-MPO multiplication might outweigh than the benifits of MPS.

Is the statement above more clear now?

>     - Similarly, shot-based ``var()`` measurements are recommended for the ``lightning.tensor`` device. The analytical calculation of ``var()`` may also result in excessive memory usage or be impractical due to prohibitive computational costs for large-scale quantum simulations.

Reviewer: Same as above, why is shot based stuff better?

Author: Good question! The updated context should be more clear.

>     - It is advisable to disable ``new_opmath`` for the ``lightning.tensor`` device, as it only supports 1-wire Hermitian observables.

Reviewer 2: ```suggestion
    - It is advisable to disable ``new_opmath`` for the ``lightning.tensor`` device, as the device only supports 1-wire Hermitian observables.
```

Reviewer 2: Just to make 100% clear that's the device and not the new_opmath option that supports only the 1-wire observables

Reviewer: same recommendation as above

Operations and observables support
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The "lightning.tensor" supports all gate operations supported by PennyLane.

Reviewer: 🙌
